### ISSUE TITLE
 UpSampling2D and StridedSlice: Keras->Tensorflow->OpenCV

### ISSUE THREAD

**Issue by gorbad-6** [opencv/opencv#10106]
I am trying to load a network which has an UpSampling2D layer from Keras with Tensorflow backend into OpenCV. In this framework, I cannot get rid of StridedSlice with the tools provided by Tensorflow. I could not find a solution in this issue tracker or the internet.

##### System information (version)

- OpenCV => 3.3.1
- Operating System / Platform => Ubuntu16.04 64 Bit
- Compiler = gcc 5.4.0
- Tensorflow = 1.4
- Keras = 2.0.8
- Pyhton 3.5

##### Detailed description

I am trying to deploy and UNet with opencv for semantic segmentation. For that I have built a toy model that has a down-sizing maxpooling and and upsizing upconv layers.

```
class MyUnet
@staticmethod
def conv(x, nf, sz=3, pad='same',  act='relu', kern='he_normal', stride=1):
        return Conv2D(nf, sz, padding=pad, activation=act, kernel_initializer=kern, strides=(stride, stride))(x)

@staticmethod
def upconv(x, nf):
        x = UpSampling2D(size=(2, 2))(x)
        x = MyUnet.conv(x, nf)
        return x

@staticmethod
def down_pass(x, nf, do_max=True):
        if do_max:
              x = MyUnet.maxp(x)
        x = MyUnet.conv(x, nf)
        x = MyUnet.conv(x, nf)
        return x

@staticmethod
def up_pass(x, concat, nf):
        up = MyUnet.upconv(x, nf)
        merge = Concatenate(axis=3)([concat, up])
        x = MyUnet.conv(merge, nf)
        x = MyUnet.conv(x, nf)
        return x

@staticmethod
def toy_unet(in_shape = (32, 32, 3)):
        inputs = Input(in_shape)
        conv1 = MyUnet.down_pass(BatchNormalization()(inputs, training=False), 8, False)
        conv2 = MyUnet.down_pass(conv1, 16)
        conv3 = MyUnet.up_pass(conv2, inputs, 16)
        conv4 = Conv2D( 3, 1, activation='sigmoid')(conv3)
        return Model( inputs, conv4 )
```

Which results in a 10-layer net. If I skip the up-pass (conv3), then using tensorflow's freeze_graph.py and graph_transform  to peel off unwanted nodes, then I can import the model in in opencv. With conv3 I get an error 

OpenCV Error: Unspecified error (Unknown layer type StridedSlice in op up_sampling2d_1/strided_slice) in populateNet,

##### Steps to reproduce
in pyhton:
```python
import os
import shutil
import numpy as np
import tensorflow as tf
import network.unet
import network.losses
import keras.backend as K
from tensorflow.python.framework import graph_io

sess = tf.Session()
K.set_session(sess)
my_net = MyUnet.toy_unet()
my_net.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['accuracy'])
data = np.random.random((10, 32, 32, 3))
labels=np.random.randint(3, size=(10, 32, 32, 3))
my_net.fit(data, labels, epochs=1, batch_size=5)
temp_export_dir = '/temp_export/'
if os.path.exists(temp_export_dir):
    shutil.rmtree(temp_export_dir)
os.makedirs(temp_export_dir)
checkpoint_prefix = os.path.join(temp_export_dir, "saved_checkpoint")
input_graph_name ='graph.pb'
with sess.graph.as_default():
    saver = tf.train.Saver()
    cp_path = saver.save(sess, checkpoint_prefix)
    graph_io.write_graph(sess.graph_def, temp_export_dir, input_graph_name)
```
in bash:
```bash
bazel-bin/tensorflow/python/tools/freeze_graph --input_graph=/temp_export/graph.pb --input_checkpoint=/temp_export/saved_checkpoint --output_node_names=conv2d_8/Sigmoid --output_graph=/temp_export/frozen_graph.pb
```
```bash
 bazel-bin/tensorflow/tools/graph_transforms/transform_graph --in_graph=/temp_export/frozen_graph.pb --out_graph=/temp_export/frozen_opt.pb --inputs=input_1 --outputs=conv2d_8/Sigmoid --transforms="fold_batch_norms fold_constants sort_by_execution_order strip_unused_nodes remove_nodes(op=Shape)"
```

(optimize for inference mangles the input shape, and i think is obsoleted through transform_graph)

Now, the output of the summary 
```bash
/tensorflow/bazel-bin/tensorflow/tools/graph_transforms/summarize_graph --in_graph=/temp_export/frozen_opt.pb 

Found 1 possible inputs: (name=input_1, type=float(1), shape=[?,32,32,3]) 
No variables spotted.
Found 1 possible outputs: (name=conv2d_8/Sigmoid, op=Sigmoid) 
Found 11751 (11.75k) const parameters, 0 (0) variable parameters, and 0 control_edges
Op types used: 23 Const, 8 BiasAdd, 8 Conv2D, 7 Relu, 2 Mul, 1 Add, 1 ConcatV2, 1 MaxPool, 1 Placeholder, 1 ResizeNearestNeighbor, 1 Sigmoid, 1 StridedSlice
```

As you can see the StridedSlice is still in the network and indeed loading it into a cv::dnn:: Net gives the error.

Issues i have looked at, but did not solve my problem:
#9755  #1029 #9459 #9536

**Comment by dkurt** [comment]
@gorbad-6, please clarify a meaning of `UpConv2D` layer mentioned in the topic. There is a deconvolution layer (transposed convolution) as a kind of trainable way to upsample feature maps: https://keras.io/layers/convolutional/#conv2dtranspose and `UpSample2D` also known as repeat-edge-padding: https://keras.io/layers/convolutional/#upsampling2d. Did you want to use a trainable upsampling or the last one?

**Comment by gorbad-6** [comment]
@dkurt I think I would like to use a way to undo MaxPooling in a controlled fashion. The second method you mentioned by UpSampling2D is the one which we chose. I will update the title as it, indeed might be confusing.

**Comment by gorbad-6** [comment]
@dkurt any hints? Sorry to be pushy, I'd be happy to try to fix this on my own, if its currently not possible with the dnn module.

**Comment by gorbad-6** [comment]
see #9353 for similar problem and putatively it has been solved in #9787, but I still struggle with it

**Comment by gorbad-6** [comment]
Solved: In the current Version of Keras, 2.0.8, the UpSamping2D layer calls the (tensorflow) backend resize_images function.

That function https://github.com/fchollet/keras/blob/master/keras/backend/tensorflow_backend.py:1798 uses for both data formats

tf.shape and slices. This is the reason that tf cannot freeze out the Shape and StridedSlice nodes. Solution: Implement your own UpSampling2D layer that handles the resize calculations without tf.shape.

**Comment by Kristufi** [comment]
Hi @gorbad-6 , would it be possible to share your solution mentioned?
Thanks.

**Comment by gorbad-6** [comment]
sure. I use the kustom keras layer
`class UpSize(Layer):
    """Reimplementation of keras.layers.UpSampling2D"""

    def __init__(self, size=(2, 2), data_format=None, **kwargs):
        """
        Constructor

        :param size: the factor by which to resize the input layer
        :param data_format: whether the net is in NHWC or NCHW format
        :param align_corners: (optional) which nearest neighbor method to use
        """
        super(UpSize, self).__init__(**kwargs)
        self.data_format = common.normalize_data_format(data_format)
        self.size = conv_utils.normalize_tuple(size, 2, 'size')
        self.input_spec = InputSpec(ndim=4)
        self.align_corners = False
        if "align_corners" in kwargs:
            self.align_corners = kwargs["align_corners"]

    def compute_output_shape(self, input_shape):
        """Computs output shape of layer"""
        if self.data_format == 'channels_first':
            height = self.size[0] * input_shape[2] if input_shape[2] is not None else None
            width = self.size[1] * input_shape[3] if input_shape[3] is not None else None
            return (input_shape[0], input_shape[1], height, width)
        elif self.data_format == 'channels_last':
            height = self.size[0] * input_shape[1] if input_shape[1] is not None else None
            width = self.size[1] * input_shape[2] if input_shape[2] is not None else None
            return (input_shape[0], height, width, input_shape[3])

    def call(self, inputs):
        """Up-samples the input

        :param inputs: 4-D array in the channels-last format
        :return x: upsampled image
        """
        if self.data_format == 'channels_last':
            height_factor, width_factor = self.size
            original_shape = int_shape(inputs)
            new_shape = tf.constant(
                np.array([original_shape[1] * height_factor,
                          original_shape[2] * width_factor]).astype('int32'))
            x = tf.image.resize_nearest_neighbor(
                inputs, new_shape, align_corners=self.align_corners)
            x.set_shape((None, original_shape[1] * height_factor if original_shape[1] is not None
                         else None, original_shape[2] * width_factor
                         if original_shape[2] is not None else None, None))
            return x
        else:
            raise ValueError('Invalid data_format:', self.data_format, ' is not supported yet.')

    def get_config(self):
        """Returns Layer configs"""
        config = {'size': self.size, 'data_format': self.data_format}
        base_config = super(UpSize, self).get_config()
        return dict(list(base_config.items()) + list(config.items()))`
